controllers/containerizedworkload_controller_helper.go

```diff
@@ -0,0 +1,158 @@
+package controllers
+
+import (
+	"context"
+	"strings"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	oamv1alpha2 "github.com/oam-dev/core-resource-controller/api/v1alpha2"
+)
+
+const (
+	OAMResourceTypeLabel = "oam.dev/type"
+	OAMResourceNameLabel = "oam.dev/name"
+	KindDeployment       = "Deployment"
+	KindService          = "Service"
+)
+
+// OAMResourceTypes
+type OAMResourceTypes string
+
+// Supported OAM Resource Types
+const (
+	workloadType OAMResourceTypes = "workload"
+)
+
+// create a corresponding deployment
+func (r *ContainerizedWorkloadReconciler) renderWorkload(ctx context.Context,
+	workload *oamv1alpha2.ContainerizedWorkload) (*appsv1.Deployment, error) {
+	var RevisionHistoryLimit int32 = 100
+	deployName := workload.Name + "-deployment"
+	depl := appsv1.Deployment{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: appsv1.SchemeGroupVersion.String(),
+			Kind:       KindDeployment,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      deployName,
+			Namespace: workload.Namespace,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{OAMResourceTypeLabel: string(workloadType), OAMResourceNameLabel: deployName},
+			},
+			RevisionHistoryLimit: &RevisionHistoryLimit,
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{OAMResourceTypeLabel: string(workloadType), OAMResourceNameLabel: deployName},
+				},
+				Spec: corev1.PodSpec{
+					Containers: workload.Spec.Containers,
```

Comment:
This is the benefit of using the k8s definition of a `Container`. Otherwise, we have to do something like this :disappointed:  https://github.com/crossplane/addon-oam-kubernetes-remote/pull/1/files#diff-16ffed2f20ffd735957e4841773fe7e6R88

Reply:
exactly... I didn't want to spend too much time on the boilerplate translation code initially especially I knew you've done it. Do you want to refactor that part out so I can resue?

Reply:
Absolutely! I think we can share quite a bit of the code between these two projects :) Generally I think the actual workload "translators" and trait "modifiers" can be used across both, and all we will need to do is swap out the generic reconcilers (i.e. workload: https://github.com/crossplane/addon-oam-kubernetes-remote/pull/1/files#diff-845bf5479da8a49c8d146d9e8157cc4e and trait: https://github.com/crossplane/addon-oam-kubernetes-remote/pull/1/files#diff-2257c722ec2c7b30a1af085a8cb1cf55). This can be a little hard to see while my PR is still in draft, so let me see if I can make it a little more concrete so we can see where we can share :+1:

Reply:
Sure, one thing I want to clarify is that the "local" controllers are meant to be generic to any OAM platform builders while the "remote" part is a crossplane specific multiple cloud implementation. Therefore, there is some degree of separation needed even if they are very similar. IMO, the OAM container -> K8s container translation code should actually belong to the generic one.

Reply:
> the "local" controllers are meant to be generic to any OAM platform builders

Are we expecting there to be multiple OAM platforms built using the Kubernetes API server as their control plane? I thought we planned to align on Crossplane as the one OSS, Kubernetes-based OAM implementation.

Reply:
I think crossplane could be the official one while other company could integrate some of our open source implementation to their internal OAM platform.

Reply:
> Are we expecting there to be multiple OAM platforms built using the Kubernetes API server as their control plane? I thought we planned to align on Crossplane as the one OSS, Kubernetes-based OAM implementation.

I think there are similar conversations happening at the governing model meetings, we can save the discussion here :)

I think there are similar conversations happening at the governing model meetings, we can save the discussion there :)